// @pulumi/kubernetes/events/v1/event.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Event = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../../utilities");
/**
 * Event is a report of an event somewhere in the cluster. It generally denotes some state change in the system. Events have a limited retention time and triggers and messages may evolve with time.  Event consumers should not rely on the timing of an event with a given Reason reflecting a consistent underlying trigger, or the continued existence of events with that Reason.  Events should be treated as informative, best-effort, supplemental data.
 */
class Event extends pulumi.CustomResource {
    /**
     * Create a Event resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (!opts.id) {
            if ((!args || args.eventTime === undefined) && !opts.urn) {
                throw new Error("Missing required property 'eventTime'");
            }
            resourceInputs["action"] = args ? args.action : undefined;
            resourceInputs["apiVersion"] = "events.k8s.io/v1";
            resourceInputs["deprecatedCount"] = args ? args.deprecatedCount : undefined;
            resourceInputs["deprecatedFirstTimestamp"] = args ? args.deprecatedFirstTimestamp : undefined;
            resourceInputs["deprecatedLastTimestamp"] = args ? args.deprecatedLastTimestamp : undefined;
            resourceInputs["deprecatedSource"] = args ? args.deprecatedSource : undefined;
            resourceInputs["eventTime"] = args ? args.eventTime : undefined;
            resourceInputs["kind"] = "Event";
            resourceInputs["metadata"] = args ? args.metadata : undefined;
            resourceInputs["note"] = args ? args.note : undefined;
            resourceInputs["reason"] = args ? args.reason : undefined;
            resourceInputs["regarding"] = args ? args.regarding : undefined;
            resourceInputs["related"] = args ? args.related : undefined;
            resourceInputs["reportingController"] = args ? args.reportingController : undefined;
            resourceInputs["reportingInstance"] = args ? args.reportingInstance : undefined;
            resourceInputs["series"] = args ? args.series : undefined;
            resourceInputs["type"] = args ? args.type : undefined;
        }
        else {
            resourceInputs["action"] = undefined /*out*/;
            resourceInputs["apiVersion"] = undefined /*out*/;
            resourceInputs["deprecatedCount"] = undefined /*out*/;
            resourceInputs["deprecatedFirstTimestamp"] = undefined /*out*/;
            resourceInputs["deprecatedLastTimestamp"] = undefined /*out*/;
            resourceInputs["deprecatedSource"] = undefined /*out*/;
            resourceInputs["eventTime"] = undefined /*out*/;
            resourceInputs["kind"] = undefined /*out*/;
            resourceInputs["metadata"] = undefined /*out*/;
            resourceInputs["note"] = undefined /*out*/;
            resourceInputs["reason"] = undefined /*out*/;
            resourceInputs["regarding"] = undefined /*out*/;
            resourceInputs["related"] = undefined /*out*/;
            resourceInputs["reportingController"] = undefined /*out*/;
            resourceInputs["reportingInstance"] = undefined /*out*/;
            resourceInputs["series"] = undefined /*out*/;
            resourceInputs["type"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const aliasOpts = { aliases: [{ type: "kubernetes:core/v1:Event" }, { type: "kubernetes:events.k8s.io/v1beta1:Event" }] };
        opts = pulumi.mergeOptions(opts, aliasOpts);
        super(Event.__pulumiType, name, resourceInputs, opts);
    }
    /**
     * Get an existing Event resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, opts) {
        return new Event(name, undefined, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Event.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Event.__pulumiType;
    }
}
exports.Event = Event;
/** @internal */
Event.__pulumiType = 'kubernetes:events.k8s.io/v1:Event';
