// @pulumi/kubernetes/provider.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Provider = void 0;
const pulumi = require("@pulumi/pulumi");
const types_1 = require("./types");
const utilities = require("./utilities");
/**
 * The provider type for the kubernetes package.
 */
class Provider extends pulumi.ProviderResource {
    /**
     * Create a Provider resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        var _a, _b, _c, _d, _e;
        let resourceInputs = {};
        opts = opts || {};
        {
            resourceInputs["cluster"] = args ? args.cluster : undefined;
            resourceInputs["context"] = args ? args.context : undefined;
            resourceInputs["enableDryRun"] = pulumi.output((_a = (args ? args.enableDryRun : undefined)) !== null && _a !== void 0 ? _a : utilities.getEnvBoolean("PULUMI_K8S_ENABLE_DRY_RUN")).apply(JSON.stringify);
            resourceInputs["enableReplaceCRD"] = pulumi.output((_b = (args ? args.enableReplaceCRD : undefined)) !== null && _b !== void 0 ? _b : utilities.getEnvBoolean("PULUMI_K8S_ENABLE_REPLACE_CRD")).apply(JSON.stringify);
            resourceInputs["helmReleaseSettings"] = pulumi.output(args ? (args.helmReleaseSettings ? pulumi.output(args.helmReleaseSettings).apply(types_1.input.helmReleaseSettingsProvideDefaults) : undefined) : undefined).apply(JSON.stringify);
            resourceInputs["kubeClientSettings"] = pulumi.output(args ? (args.kubeClientSettings ? pulumi.output(args.kubeClientSettings).apply(types_1.input.kubeClientSettingsProvideDefaults) : undefined) : undefined).apply(JSON.stringify);
            resourceInputs["kubeconfig"] = (_c = (args ? args.kubeconfig : undefined)) !== null && _c !== void 0 ? _c : utilities.getEnv("KUBECONFIG");
            resourceInputs["namespace"] = args ? args.namespace : undefined;
            resourceInputs["renderYamlToDirectory"] = args ? args.renderYamlToDirectory : undefined;
            resourceInputs["suppressDeprecationWarnings"] = pulumi.output((_d = (args ? args.suppressDeprecationWarnings : undefined)) !== null && _d !== void 0 ? _d : utilities.getEnvBoolean("PULUMI_K8S_SUPPRESS_DEPRECATION_WARNINGS")).apply(JSON.stringify);
            resourceInputs["suppressHelmHookWarnings"] = pulumi.output((_e = (args ? args.suppressHelmHookWarnings : undefined)) !== null && _e !== void 0 ? _e : utilities.getEnvBoolean("PULUMI_K8S_SUPPRESS_HELM_HOOK_WARNINGS")).apply(JSON.stringify);
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(Provider.__pulumiType, name, resourceInputs, opts);
    }
    /**
     * Returns true if the given object is an instance of Provider.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Provider.__pulumiType;
    }
}
exports.Provider = Provider;
/** @internal */
Provider.__pulumiType = 'kubernetes';
