// @pulumi/kubernetes/rbac/v1beta1/roleBinding.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoleBinding = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../../utilities");
/**
 * RoleBinding references a role, but does not contain it.  It can reference a Role in the same namespace or a ClusterRole in the global namespace. It adds who information via Subjects and namespace information by which namespace it exists in.  RoleBindings in a given namespace only have effect in that namespace. Deprecated in v1.17 in favor of rbac.authorization.k8s.io/v1 RoleBinding, and will no longer be served in v1.20.
 */
class RoleBinding extends pulumi.CustomResource {
    /**
     * Create a RoleBinding resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (!opts.id) {
            if ((!args || args.roleRef === undefined) && !opts.urn) {
                throw new Error("Missing required property 'roleRef'");
            }
            resourceInputs["apiVersion"] = "rbac.authorization.k8s.io/v1beta1";
            resourceInputs["kind"] = "RoleBinding";
            resourceInputs["metadata"] = args ? args.metadata : undefined;
            resourceInputs["roleRef"] = args ? args.roleRef : undefined;
            resourceInputs["subjects"] = args ? args.subjects : undefined;
        }
        else {
            resourceInputs["apiVersion"] = undefined /*out*/;
            resourceInputs["kind"] = undefined /*out*/;
            resourceInputs["metadata"] = undefined /*out*/;
            resourceInputs["roleRef"] = undefined /*out*/;
            resourceInputs["subjects"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const aliasOpts = { aliases: [{ type: "kubernetes:rbac.authorization.k8s.io/v1:RoleBinding" }, { type: "kubernetes:rbac.authorization.k8s.io/v1alpha1:RoleBinding" }] };
        opts = pulumi.mergeOptions(opts, aliasOpts);
        super(RoleBinding.__pulumiType, name, resourceInputs, opts);
    }
    /**
     * Get an existing RoleBinding resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, opts) {
        return new RoleBinding(name, undefined, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of RoleBinding.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === RoleBinding.__pulumiType;
    }
}
exports.RoleBinding = RoleBinding;
/** @internal */
RoleBinding.__pulumiType = 'kubernetes:rbac.authorization.k8s.io/v1beta1:RoleBinding';
